File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/platform/operation/PlatformResult.java
package top.likoslupus.cellulosesz.api.platform.operation;

import java.util.Optional;

import static top.likoslupus.cellulosesz.api.validation.TextChecks.requireNonBlank;

import static java.util.Objects.requireNonNull;

/** Strongly constrained platform operation outcome. */
public final class PlatformResult<T> {

    private final PlatformOperationStatus status;
    private final Optional<T> value;
    private final String detail;

    private PlatformResult(
            PlatformOperationStatus status,
            Optional<T> value,
            String detail
    ) {
        this.status = requireNonNull(status, "status");
        this.value = requireNonNull(value, "value");
        this.detail = requireNonNull(detail, "detail");
        var successful = status == PlatformOperationStatus.SUCCESS
                || status == PlatformOperationStatus.PARTIAL_SUCCESS;
        if (!successful && value.isPresent()) {
            throw new IllegalArgumentException("Failure results cannot carry a value");
        }
        if (status == PlatformOperationStatus.PARTIAL_SUCCESS) {
            requireNonBlank(detail, "detail");
        } else if (status == PlatformOperationStatus.SUCCESS && !detail.isEmpty()) {
            throw new IllegalArgumentException("Successful results cannot carry failure detail");
        } else if (!successful) {
            requireNonBlank(detail, "detail");
        }
    }

    public static <T> PlatformResult<T> success(T value) {
        return new PlatformResult<>(
                PlatformOperationStatus.SUCCESS,
                Optional.of(requireNonNull(value, "value")),
                ""
        );
    }

    public static PlatformResult<Void> success() {
        return new PlatformResult<>(PlatformOperationStatus.SUCCESS, Optional.empty(), "");
    }

    public static <T> PlatformResult<T> partial(T value, String detail) {
        return new PlatformResult<>(
                PlatformOperationStatus.PARTIAL_SUCCESS,
                Optional.of(requireNonNull(value, "value")),
                detail
        );
    }

    public static <T> PlatformResult<T> failure(PlatformOperationStatus status, String detail) {
        if (status == PlatformOperationStatus.SUCCESS
                || status == PlatformOperationStatus.PARTIAL_SUCCESS
        ) {
            throw new IllegalArgumentException(
                    "Use success or partial factories for successful results");
        }
        return new PlatformResult<>(status, Optional.empty(), detail);
    }

    public PlatformOperationStatus status() {
        return status;
    }

    public Optional<T> value() {
        return value;
    }

    public String detail() {
        return detail;
    }

    public boolean successful() {
        return status == PlatformOperationStatus.SUCCESS
                || status == PlatformOperationStatus.PARTIAL_SUCCESS;
    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/platform/operation/PlatformResult.kt
package top.likoslupus.cellulosesz.api.platform.operation

import top.likoslupus.cellulosesz.api.validation.requireNonBlank

/** Strongly constrained platform operation outcome. */
public class PlatformResult<T> private constructor(
    private val status: PlatformOperationStatus,
    private val rawValue: T?,
    private val detail: String
) {

    init {
        val successful = status == PlatformOperationStatus.SUCCESS ||
                status == PlatformOperationStatus.PARTIAL_SUCCESS
        require(!(!successful && rawValue != null)) { "Failure results cannot carry a value" }
        if (status == PlatformOperationStatus.PARTIAL_SUCCESS) {
            detail.requireNonBlank { "detail" }
        } else require(!(status == PlatformOperationStatus.SUCCESS && detail.isNotEmpty())) {
            "Successful results cannot carry failure detail"
        }
        if (!successful) {
            detail.requireNonBlank { "detail" }
        }
    }

    public fun status(): PlatformOperationStatus = status

    public fun value(): T? = rawValue

    public fun detail(): String = detail

    public fun successful(): Boolean =
        status == PlatformOperationStatus.SUCCESS ||
                status == PlatformOperationStatus.PARTIAL_SUCCESS

    public companion object {

        @JvmStatic
        public fun <T> success(value: T): PlatformResult<T> =
            PlatformResult(
                PlatformOperationStatus.SUCCESS,
                requireNotNull(value) { "value" },
                ""
            )

        @JvmStatic
        public fun success(): PlatformResult<Void> =
            PlatformResult(PlatformOperationStatus.SUCCESS, null, "")

        @JvmStatic
        public fun <T> partial(value: T, detail: String): PlatformResult<T> =
            PlatformResult(
                PlatformOperationStatus.PARTIAL_SUCCESS,
                requireNotNull(value) { "value" },
                detail
            )

        @JvmStatic
        public fun <T> failure(status: PlatformOperationStatus, detail: String): PlatformResult<T> {
            require(
                status != PlatformOperationStatus.SUCCESS &&
                        status != PlatformOperationStatus.PARTIAL_SUCCESS
            ) {
                "Use success or partial factories for successful results"
            }
            return PlatformResult(status, null, detail)
        }

    }

}
